### PR TITLE
Add catalog to minio

### PR DESCRIPTION
**Pull request by Zyqsempai**
**What is the problem I am trying to address?**

Implementing Catalog method into minio storage package

**Mention the issue number it fixes or add the details of the changes if it doesn't have a specific issue.**

Fixes https://github.com/gomods/athens/issues/996

### CODE REVIEW

**Comment by manugupt1** [comment]
Minio tests are failing, can you fix that please?

File: pkg/storage/minio/minio_test.go [hunk 1]
```diff
@@ -19,10 +19,8 @@ func BenchmarkBackend(b *testing.B) {
 }
 
 func (s *storageImpl) clear() error {
-	doneCh := make(chan struct{})
-	defer close(doneCh)
-	objectCh := s.minioClient.ListObjectsV2(s.bucketName, "", true, doneCh)
-	for object := range objectCh {
+	objectCh, _ := s.minioCore.ListObjectsV2(s.bucketName, "", "", false,"",0,"")
```
Comment (manugupt1) [inline pkg/storage/minio/minio_test.go]: Is there a reason to make this change?

File: pkg/storage/minio/minio.go [hunk 2]
```diff
@@ -12,6 +12,7 @@ import (
 
 type storageImpl struct {
 	minioClient *minio.Client
+	minioCore *minio.Core
```
Comment (manugupt1) [inline pkg/storage/minio/minio.go]: Is there a reason to switch from minio.Client to minio.Core when both of them can use ListObjectsV2?

**Comment by Zyqsempai** [comment]
@manugupt1 Fixed.

File: pkg/storage/minio/cataloger.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+// Catalog implements the (./pkg/storage).Cataloger interface
+// It returns a list of modules and versions contained in the storage
+func (s *storageImpl) Catalog(ctx context.Context, token string, pageSize int) ([]paths.AllPathParams, string, error) {
+	const op errors.Op = "minio.Catalog"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+	queryToken := token
+	res := make([]paths.AllPathParams, 0)
+	doneCh := make(chan struct{})
+	defer close(doneCh)
```
Comment (arschles) [inline pkg/storage/minio/cataloger.go]: @Zyqsempai I think this is unused. Did I miss it?

Comment (Zyqsempai) [reply pkg/storage/minio/cataloger.go]: Oh, you'r right;)

**Comment by Zyqsempai** [comment]
@arschles Done.

**Comment by arschles** [comment]
@Zyqsempai @marpio pointed out https://github.com/gomods/athens/issues/996#issuecomment-453869584 to me just now. I'm wondering if you think https://github.com/gomods/athens/issues/996#issuecomment-453869584 is a problem and we should not merge because of it? Otherwise, we could follow-up with a fix to the performance.

**Comment by Zyqsempai** [comment]
@arschles For sure it would be better to check it one more time, on another side than mine, maybe it's just a problem with my local ENV, but who knows.

**Comment by Zyqsempai** [comment]
Let's merge it then;)

**Comment by arschles** [comment]
@Zyqsempai you got it! :shipit: 🎉 

Thanks so much for this. I see some of your other PRs,

**Pull request merged**